source/algorithms.tex

```diff
@@ -11580,6 +11581,138 @@
 \end{codeblock}
 \end{itemdescr}
 
+\rSec1[alg.rand]{Specialized \tcode{<random>} algorithms}
+
+\rSec2[alg.rand.general]{General}
+
+\pnum
+The contents specified in \ref{alg.rand}
+are declared in the header \libheaderrefx{random}{rand.synopsis}.
+
+\rSec2[alg.rand.generate]{\tcode{generate_random}}
+
+\begin{itemdecl}
+template<class R, class G>
+  requires @\libconcept{output_range}@<R, invoke_result_t<G&>> && @\libconcept{uniform_random_bit_generator}@<remove_cvref_t<G>>
+constexpr borrowed_iterator_t<R> ranges::generate_random(R&& r, G&& g);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+\begin{itemize}
+\item
+Calls \tcode{g.generate_random(std::forward<R>(r))}
+if this expression is well-formed.
+\item
+Otherwise, if \tcode{R} models \tcode{sized_range},
+fills \tcode{r} with \tcode{ranges::size(r)} values of
+type \tcode{invoke_result_t<G\&>} by performing
+an unspecified number of invocations of
+the form \tcode{g()} or \tcode{g.generate_random(s)},
+if such an expression is well-formed for a value \tcode{N} and
+an object \tcode{s} of type \tcode{span<invoke_result_t<G\&>, N>}.
+\begin{note}
+Values of \tcode{N} can differ between invocations.
+\end{note}
+\item
+Otherwise, calls \tcode{ranges::generate(std::forward<R>(r), ref(g))}.
+\end{itemize}
+
+\pnum
+\returns
+\tcode{ranges::end(r)}.
+
+\pnum
+\remarks
+The effects of \tcode{generate_random(r, g)} shall be equivalent to
+\tcode{ranges::generate(std::for\-ward<R>(r), ref(g))}.
```

Comment:
Should we turn this into _Preconditions_?

Reply:
Uh, why? This is a constraint on the implementation, not the user.

Reply:
`ranges::generate_random` potentially calls a `generate_random` member function, and the `generate_random` member function is only provided by the user code (at this moment). In such a case, the equivalency can't be achieved if the `generate_random` member function doesn't have equivalent effects.

Reply:
Right, this is a requirement on user code

Reply:
... But only if the generator has a suitable member function that gets used here.

Reply:
Ok.  So, going back to the original concern here: should this be a _Precondition_ instead?

Reply:
@jwakely any thoughts?

Reply:
Yes, but maybe we should do that as an LWG issue, since I think it needs something non-trivial like "if the first or second bullet below is used, then the behaviour is equivalent to the third bullet" ... but in proper standardese.

Reply:
OK -- the new WD is in the mailing, so please feel free to file issues against N4981!